### ISSUE TITLE
Feature request: sloppy release for layer keys

### ISSUE THREAD

**Issue by fleutot** [kmonad/kmonad#910]
I am having a hard time actually releasing a key in a layer, and making sure it is entirely released before I release the holding key. I am using `tap-hold-next-release`.

I understand that it is complicated to make both roll over and a more confortable layering work well together. What I would appreciate, is the ability to create this behavior:

```
(tap-hold-next-release-sloppy 100 50 a lshft)
Pa  Pb  Ra 49 Rb  -> B
Pa  Pb  Ra 51 Rb  -> ab
```
In other words, the ability to allow for an out-of-order release sequence within a time limit, and interpret that as the "other order" release sequence, in terms of interpret this as a hold anyway.

Even if I tend to release the holding key too early, I think I usually do so much closer to the modified key, than when I roll over.

Is there a way to do this already? If not, could this be considered?

**Comment by jokesper** [comment]
I assume this is bound on the button `a`.

Are you sure this is the order you would want, I think it should be the other way around. I.e. when rolling there is a short delay between presses and between releases. While holding I could see the hold button also being held longer.

**Comment by fleutot** [comment]
I think I got the order right, but happy to be proven wrong.

My hypothesis is that the time between the two releases is much shorter when I use `a` as mod, than when I roll `a` then `b`. The idea was to use this to avoid deciding that the sequence is a roll, if the releases are in wrong order for a hold, but close enough.

I'm not sure that a required delay between Pa and Pb would help me much, I think my timing between Pa and Pb is pretty similar whether I mod or roll.

Thinking of this a bit more, another algorithm to determine hold or roll could be the ratio between overlap time and total sequence time.

Pa Pb Ra Rb

-> If the overlap pressed time (between Pb and Ra) is more than X % of the total sequence time (between Pa and Rb), the sequence is probably meant as a hold, not a roll. Would it be better?

**Comment by jokesper** [comment]
Hmmm, overlap percentage is something which sounds plausible to me.
Though how do we calculate it.
If we have:
```
Pa Pb Ra Rb
\------/
   \------/
```
How would we calculate the overlap?
Overlap time devided by total time?
that would result in non 100% when we have
`Pa Tb Ra`.

**Comment by fleutot** [comment]
Maybe the overlap percentage is only relevant in a P/R sequence that would be a roll?

**Comment by fleutot** [comment]
I suppose this coiuld only be done with experience, but I see possible alternatives:

Overlap time / (overlap time + pre-hold time)
, that is (Ra - Pb) / (Ra - Pa)

Overlap time / (overlap time + post-hold time)
, that is (Ra - Pb) / (Rb - Pb)

Overlap time / sequence time
, that is (Ra - Pb) / (Rb - Pa)

The second one is akin to my original post, but any of three would be interesting to get a feel for.

**Comment by jokesper** [comment]
I will try to hack something together later and then post it here

**Comment by fleutot** [comment]
The pre-hold time, could actually be a bad thing to take into account. I don't know how other people do it, but now that I'm learning my layers, I do tend to hold a mod (in this case, a) for some undetermined time before I remember which target key. So I still think that the second one might be the preferable variant.

**Comment by jokesper** [comment]
Hmmm, I see the point, though the extreme case is not a problem since you would probably use a hold variant:
I.e. you would use `tap-hold-overlap` instead of `tap-overlap`.

**Comment by jokesper** [comment]
I just noticed a problem while implementing: when multiple buttons are pressed while one is hold. This makes it a bit more tricky.

**Comment by fleutot** [comment]
The algorithm is only relevant for R events directly following the mod's R event (as in no corresponding P in between). So for that second case, it's really about where Ra occurred between Pb and Rb.

I suppose you make that decision on Rb? So on Rb, check the time since Pb, and if there's an Ra in between, do the ratio calculation? And do that for any R? 

I understand that it's probably more complex than this. Maybe just do a proof of concept that doesn't support multiple target keys, before you put too much effort into it? Just to validate that it's a good thing before doing it for real.

**Comment by jokesper** [comment]
[I have implemented](https://github.com/jokesper/kmonad/tree/feat-tap-overlap) the first variant (Ra-Pb)/(Ra-Pa).
With completely surrounded buttons also resulting in hold. I.e. a "sloppy" tap-hold-next-release.
Personally, unless I set it to really high (like 95), I press a lot of modifiers accidentally.
I'm not quite sure if this helps me in any way. Also this version currently has a major bug: holding via the overlap check, results in a sequence of taps (`Ta Tb`) instead of the arounding behavior (`Pa Tb Ra`).

I think I now understand the reason why you would want the second case:
`Pa 100 Pb 50 Rb Ra` and
`Pa 100 Pb 50 Ra Rb` are almost the exact same, but the later results in the wrong keystrokes when using tap-hold-next-release.

**Comment by fleutot** [comment]
Yes, I agree, I think the sequence timing before both buttons are down is going to be inconsistent. I, for one, think I have pretty much the same timing on pressing down, whether I am modding or rolling. So the first variant should be all over the place. The hypothesis (which needs verifying) is that the timing of releases might make the roll distinguishable from the mod.

As you understand, that's why the second case, which entirely removes the pre-hold from the equation, looked the most interesting to me. 

As for implementation, or verification of the hypothesis, we could test to push the Ra event by a parameterizable delay, if there was a P? between Pa and Ra. I think that this would be simpler to implement?

**Comment by jokesper** [comment]
I have implemented version 2 in #910, feel free to take a look and or test it

**Comment by fleutot** [comment]
I suppose you mean to link [this](https://github.com/jokesper/kmonad/tree/feat-tap-overlap) instead.

This looks promising!

I tried to merge it with [pull-524-rebased](https://github.com/jokesper/kmonad/tree/pull-524-rebased), but the result seems to freeze after typing keys that also are user defined buttons. It's not necessarily surprising that the changes are not compatible, but if it is to you, I wanted to let you know.

**Comment by fleutot** [comment]
This is definitely an improvement for me. I start wondering if the "add a customizable delay to Ra" wouldn't be preferable. The difference in release timing is more an absolute, than dependent on how long the press overlap was before that. What do you think?

**Comment by jokesper** [comment]
> I tried to merge it with [pull-524-rebased](https://github.com/jokesper/kmonad/tree/pull-524-rebased), but the result seems to freeze after typing keys that also are user defined buttons. It's not necessarily surprising that the changes are not compatible, but if it is to you, I wanted to let you know.
Yes, or just #916.

Actually I was testing those changes together.
You should change the following:
```diff
diff --git a/src/KMonad/Model/Button.hs b/src/KMonad/Model/Button.hs
index 6705bd9..e9bfda1 100644
--- a/src/KMonad/Model/Button.hs
+++ b/src/KMonad/Model/Button.hs
@@ -466,7 +466,7 @@ tapOverlap' ms pc t h mtb = onPress' t $ do
   -- relative times, how long a key was pressed without overlap to calculate
   -- the percentage.
   whilePressed :: MonadK m => Milliseconds -> HashMap Keycode Milliseconds -> m ()
-  whilePressed time ks = register InputHook . Hook (onTimeout time) $ \r -> do
+  whilePressed time ks = register InputHookPrio . Hook (onTimeout time) $ \r -> do
     p <- matchMy Release
     let e = r^.event
     let isRel = isRelease e
@@ -486,7 +486,7 @@ tapOverlap' ms pc t h mtb = onPress' t $ do
   whileReleased :: MonadK m => Milliseconds -> Milliseconds -> HashMap Keycode Milliseconds -> m ()
   whileReleased maxMs time ks
     | null ks = doTap
-    | otherwise = tHookF InputHook (calcMaxMs ks) doTap $ \r -> do
+    | otherwise = tHookF InputHookPrio (calcMaxMs ks) doTap $ \r -> do
         let e = r^.event
         let k = e^.keycode
         let time' = time + r^.elapsed
```

**Comment by fleutot** [comment]
Awesome, that works. Not sure why th merge did not tak care of this, but as I said eariler, I don't know haskell. This will make it easier for me to evaluate the overlap.

I do sometimes get stuck in a layer, as though the release event of the layer button was lost. I cannot exit the layer by pressing the button again, I have to kill kmonad. Is it something known, or new to these branches? Have you seen anything like it?

**Comment by jokesper** [comment]
> Not sure why the merge did not take care of this, [...]

This area did not contain conflicts detectable by git. The conflict was in Internal API usage.

> I do sometimes get stuck in a layer, as though the release event of the layer button was lost.

Yes

> I cannot exit the layer by pressing the button again, I have to kill kmonad.

but not this.

I assume you did not have this issue before using my branch. How do you switch your layers?

**Comment by fleutot** [comment]
Correct, but on the other hand I am quite new to kmonad, so I'm quicker now than I was a couple of weeks ago. But without both branches you made for me, I cannot be fast, so hard to test tht in main.

**Comment by jokesper** [comment]
> You also experience loss of release event?

Yeah, though very infrequent. Though if this were our problem, it may be in the c bits.

I always ignored them since they were so, so rare, that reproducing them was virtually impossible.
Also it is not a KMonad specific problem exactly. Some programs also sometimes have this problem in combination with
tty / focus switching and or locking.

Could you send a log of when it happens. Since it's rare you may want to passively record it in a file
and restart kmonad if you noticed it happening. Then you could simply send the last few events or so (maybe like 100~200 lines).

**Comment by fleutot** [comment]
> Could you send a log of when it happens

[Here goes](https://github.com/kmonad/kmonad/issues/922), although it was not an a layer-toggle button, and I was able to reset it by repressing. We can assume that it is the same that happened on my navigation layer, but I got confused because I have multiple buttons toggling that layer.

**Comment by jokesper** [comment]
> Any chance you might build a test that only makes the Ra pending for while (constant), in case Rb happens later?

Do you mean specifying an absolute time the release may come early instead of a relative time.
If so then this would be far easier to implement (vell I already did the complicated part but it's also about complexity when reading the code)

**Comment by jokesper** [comment]
I updated #916 for this. since this replaces tap-hold-next-release you may get a large conflict when merging. I would simply use the new version when resolving and change the inputhook to use the prio suffix

**Comment by fleutot** [comment]
Thank you. I didn't manage to fix the InputHook after [merge](https://github.com/fleutot/kmonad/commit/54b79e41449c8a34b1454830f0054ff20b9d0cd7#r149299627), though. Here is my [attempt](https://github.com/fleutot/kmonad/commit/5d318f4fadecd00379024f0203b2e8b47aa7b921#r149299582) to fix [Button.hs](https://github.com/fleutot/kmonad/blob/5d318f4fadecd00379024f0203b2e8b47aa7b921/src/KMonad/Model/Button.hs#L379), could you please assist? I'm getting no output on single keys:

```
Running hooks
Storing event, current store: 
 - WrappedKeyEvent Release <t>, NoCatch
 - WrappedKeyEvent Press <t>, NoCatch
 - WrappedKeyEvent Release <i>, NoCatch
 - WrappedKeyEvent Press <i>, NoCatch
 - WrappedKeyEvent Release <g>, NoCatch
 - WrappedKeyEvent Press <g>, NoCatch
 - WrappedKeyEvent Release <f>, NoCatch
 - WrappedKeyEvent Press <f>, NoCatch
 - WrappedKeyEvent Release <f>, NoCatch
 - WrappedKeyEvent Press <f>, NoCatch
 - WrappedKeyEvent Release <f>, NoCatch
```

**Comment by jokesper** [comment]
> Should I replace all instances of InputHook to InputHookPrio? (Testing this locally as we speak, but would like to know if that is correct.)

no, but you missed the other important part in line 399.

**Comment by jokesper** [comment]
> I would have appreciated a new branch for this, since it is an alternative solution

The problem is, that I can't change the branch of pull requests. Maybe I will have two branches:
1. the pull request
2. the testing

With the exact same commit, but then I can only rebase / ammend the first.

> A nice extra (I am just dreaming here) would be for the debug output to print the interval between the releases, that made it decide if the sequence was a roll or a hold.

Agreed. Our current logging is very confusing to users, and developers. Printing the time delay, (maybe decision making), and something like hook labels (also deduplicate rehooked ids) would really help. It's on my todo list.

> With an output format that would let people parse and plot

maybe a `--json` flag would go a long way.

**Comment by fleutot** [comment]
> The problem is, that I can't change the branch of pull requests.

So, don't make a pull request before the feature is ready for merge? Keep a branch while developing and testing, I can fetch that branch without a PR. Only make a PR when you consider the work done. You can also alway decline your own PR, if you don't want to request a merge for that work anymore. Personally, I don't care much for amends and rebases anyway, I think there's no real value in changing or hiding history, but I know that many prefer a tidy history rather than an actual history.

> maybe a --json flag would go a long way.
Maybe. What first popped into my mind was parsing the text file and plotting with python and matplotlib. I think the lines of dashes in the debug output are useful for that type of parsing, easy to parse a group of data lines that fit together.

> no, but you missed the other important part in line 399.
Ok, thanks! At that point I am just taking chances. But I've been running a couple of days with all 11 instances of `InputHook` replaces with `InputHookPrio`, and saw no drawback. Anyway, I'm running now with only these two instances modified. And amended that commit, contradicting myself, since this is just a temporary commit for test.

**Comment by jokesper** [comment]
> So, don't make a pull request before the feature is ready for merge?

Hmm, true.

> Only make a PR when you consider the work done

I guess I was to hasty on that, since even though it was a solution it wasn't the ideal one.

> But I've been running a couple of days with all 11 instances of InputHook replaced with InputHookPrio, and saw no drawback

This may be due to you not using those buttons in combination with the tap-hold buttons.

effectily you disabled the `hold` behaviour for all complex buttons (partially).

The parts where it was listed in `Buttons.hs` explicitly are all called when a new button is pressed, and therefore should not really matter. Currently there is no situation where this would be relevant, though I wouldn't change it this way in master.
